source/tdx-guest-hardening.rst

```diff
@@ -1000,4 +1000,325 @@ https://github.com/intel/ccc-linux-guest-hardening/tree/master/bkc/kafl/userspac
 
+Enabling additional kernel drivers
+==================================
 
+The reference TDX guest kernel implementation provided for the `Linux SW stack for
+Intel TDX <https://github.com/intel/tdx-tools>`_ only enables a small set of
+virtio drivers that are essential for the TDX guest basic functionality. These
+drivers have been hardened using the methodology described in this document,
+but naturally different deployment scenarios and use cases for the TDX will
+require many more additional drivers to be enabled in the TDX guest kernel.
+
+This section provides guidance on how to use the methodology presented
+in this document for adding and hardening a new driver for the TDX guest kernel.
+
+Step-by-step guidance
+---------------------
+
+In order to explain better on how to perform the below steps, we will
+use virtio-vsock driver as an example. This driver was the last one to
+be enabled and hardened
+for the `Linux SW stack for Intel TDX <https://github.com/intel/tdx-tools>`_.
+Its primary usage in TDX guest kernel is to communicate with the host to
+request converting a local TDX attestation report into a remotely verifiable
+TDX attestation quote.
+
+1. **Identify the device/driver pair**. The first step includes
+locating the source code of a target driver in the Linux kernel tree,
+understanding the bus that this driver
+is registered for (typically it would be a pci or acpi bus), as well as
+how the driver registration is done, how to perform functional testing
+for this driver and any higher-level interface abstractions present.
+
+**Example**. For our virtio-vsock example, the source code of this
+driver is located at /net/vmw_vsock/virtio_transport.c and the driver
+registers itself on the virtio bus (an abstraction level over the pci bus)
+using register_virtio_driver().
+
+2. **Perform code audit**. In this step, the source code of the driver
+is manually audited to determine the input points where the untrusted data
+from the host/VMM is consumed and how this data is being processed.
+In order to facilitate the manual audit, the check_host_input smatch pattern can
+be used to identify these input points. For that, a smatch run can be done on
+an individual driver source file using kchecker command.
+
+**Example**. The below command line for virtio-vsock driver assumes
+that you have
+a smatch instance with the check_host_input pattern installed at
+~/smatch folder and
+the command is invoked from the kernel source tree root.
+For the instructions on how to install smatch please consult
+`README.md <https://github.com/intel/ccc-linux-guest-hardening/blob/master/bkc/audit/README.md>`_
+
+   .. code-block:: bash
+
+      ~/smatch_scripts/kchecker net/vmw_vsock/virtio_transport.c > driver_results
+
+The driver_results output file will contain the list of input points
+and the limited
+propagation information:
+
+   .. code-block:: shell
```

Comment:
The `.. code-block:: shell` directive shouldn't be indented.
That's why the parsing failed here and the output looses the line breaks:
![image](https://user-images.githubusercontent.com/964610/178297938-27bb6ccf-ea16-4067-a9d8-28b92de479d7.png)


Instead, by removing the indentation like I suggest:
![image](https://user-images.githubusercontent.com/964610/178298021-067a79b3-4fc0-4788-bb73-d669b9c517d9.png)

Results in:
![image](https://user-images.githubusercontent.com/964610/178298083-93e66ebe-e06c-4315-b200-4e5c2963e119.png)